snuba/pipeline/plans_selector.py

```diff
@@ -42,17 +41,16 @@ def select_best_plans(
     all the selected plans are in the same storage set and that the sum
     of the ranking is minimal.
     """
+    plan_data = [
+        [PlanData(rank, alias, plan) for rank, plan in enumerate(_plans)]
+        for alias, _plans in plans.items()
+    ]
 
-    ranks: MutableMapping[StorageSetKey, StorageSetPlansCandidate] = {}
-    for alias, viable_plans in plans.items():
-        for rank, plan in enumerate(viable_plans):
-            if plan.storage_set_key not in ranks:
-                ranks[plan.storage_set_key] = StorageSetPlansCandidate()
-            ranks[plan.storage_set_key].add_plan(alias, plan, rank)
+    all = [PlanCandidate(data) for data in itertools.product(*plan_data)]
 
     candidates = sorted(
-        [r for r in ranks.values() if r.is_complete(len(plans))],
-        key=lambda candidate: candidate.total_cost,
+        [candidate for candidate in all if candidate.is_valid()],
+        key=lambda candidate: candidate.get_rank_sum(),
```

Comment:
Why does this change from having a cost metrics to simply comparing the sum of the rankings? 
While in the current implementation uses the rank as cost function, the cost is arbitrary, that would have allowed us to use the same data structures even if we introduced a more complex and meaningful cost calculation that took the cost of joining two specific table into account.
Not pushing back on this implementation as it is effectively equivalent to the previous one in terms of results.

Reply:
I tried not to get too deep into the cost calculation logic while keeping this functionally the same as before. It took me quite a long time to understand this (I suspect in part because there is a lot of complexity to support stuff that doesn't exist yet) and I think I may have subconsciously attempted to simplify a little bit.